sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +26,167 @@ template <typename T, access::address_space Space> class multi_ptr;
 
 namespace detail {
 
-template <typename> struct is_vec : std::false_type {};
-template <typename T, std::size_t N>
-struct is_vec<cl::sycl::vec<T, N>> : std::true_type {};
+namespace sub_group {
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_signed<T>::value, T>::type
-calc(T x, intel::minimum<T> op) {
-  return __spirv_GroupSMin(__spv::Scope::Subgroup, O, x);
+template <typename T> T broadcast(T x, id<1> local_id) {
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;
+  OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);
+  OCLTy Ret = __spirv_GroupBroadcast(__spv::Scope::Subgroup, Arg,
+                                            local_id.get(0));
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_unsigned<T>::value, T>::type
-calc(T x, intel::minimum<T> op) {
-  return __spirv_GroupUMin(__spv::Scope::Subgroup, O, x);
+template <typename T> T shuffle(T x, id<1> local_id) {
```

Comment:
I like that these are no longer limited to arithmetic types.  Should it still check to see whether `T` can be stored in an OpenCL type?

Reply:
Actually, these functions wasn't limited to arithmetic types from the beginning.

Regarding the check: I'm not sure that this is safe, because of two reasons:
- [SPV_INTEL_subgroups](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/INTEL/SPV_INTEL_subgroups.html) requires arguments   to be scalar or vector types
- [cl_intel_subgroups](https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html) restrict list of types which can be used in this built-in

I think for now it is better to stick with types which are compatible with OpenCL ones. As further extension/improvement, we could support arbitrary `trivially_copyable` types by performing several shuffles on a chunk of data

Reply:
They weren't limited to arithmetic types by the specification, but I thought they were in the original implementation.

Either way, I agree we should move towards supporting arbitrary `trivially_copyable` types.